src/scitokens/utils/keycache.py

```diff
@@ -233,26 +234,49 @@ def _get_issuer_publickey(issuer, key_id=None, insecure=False):
         headers={'User-Agent' : 'SciTokens/{}'.format(PKG_VERSION)}
 
         # Go to the issuer's website, and download the OAuth well known bits
-        # https://tools.ietf.org/html/draft-ietf-oauth-discovery-07
-        well_known_uri = ".well-known/openid-configuration"
-        if not issuer.endswith("/"):
-            issuer = issuer + "/"
-        parsed_url = urlparse.urlparse(issuer)
-        updated_url = urlparse.urljoin(parsed_url.path, well_known_uri)
-        parsed_url_list = list(parsed_url)
-        parsed_url_list[2] = updated_url
-        meta_uri = urlparse.urlunparse(parsed_url_list)
-
-        # Make sure the protocol is https
-        if not insecure:
-            parsed_url = urlparse.urlparse(meta_uri)
-            if parsed_url.scheme != "https":
-                raise NonHTTPSIssuer("Issuer is not over HTTPS.  RFC requires it to be over HTTPS")
-        response = request.urlopen(request.Request(meta_uri, headers=headers))
-        data = json.loads(response.read().decode('utf-8'))
+        # https://tools.ietf.org/html/rfc8414
+        # We first try the RFC's metadata location, then try the openid defined
```

Comment:
Unfortunately, RFC8414 is more complex.  To quote:

> The client would make the following request when the
>   issuer identifier is "https://example.com/issuer1" and the well-known
>   URI suffix is "oauth-authorization-server" to obtain the metadata,
>   since the issuer identifier contains a path component:
>
>     GET /.well-known/oauth-authorization-server/issuer1 HTTP/1.1
>     Host: example.com
>
>   Using path components enables supporting multiple issuers per host.
>   This is required in some multi-tenant hosting configurations.  This
>   use of ".well-known" is for supporting multiple issuers per host;
>   unlike its use in RFC 5785 [RFC5785], it does not provide general
>   information about the host.

So for the RFC8414 case, you have to prepend `.well-known/oauth-authorization-server` to the path while in OIDC you add it as a suffix.

Can you also make the ordering of the discovery (OIDC or OAuth2) something specified in the configuration file instead of embedded in the code?